File: Old Code/WurmRancher/WurmRancher/Wurm.cs
using System;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;

namespace WurmRacher
{
    public class Wurm 
    {
        WurmHead head;
        BackAttachable tail;
        int length;
        GameControl theControl;

        int food_eaten_since_growth = 0;
        const int food_per_segment = 4;

        public event EventHandler<GameEventArgs> Eats;
        public event EventHandler<GameEventArgs> LengthChange;

        public Wurm(GameControl theControl_, int length_, Point starting_pos)
        {
            theControl = theControl_;
            length = length_;
            head = new WurmHead(theControl);
            theControl.AddCreature(head, starting_pos);
            WurmBodyPiece w2 = new WurmBodyPiece(theControl, head, head);
            w2.EatenByParasite += new EventHandler<EventArgs>(PieceEatenByParasite);
            theControl.AddCreature(w2, starting_pos);

            for (int i = 0; i < length - 2; i++)
            {
                WurmBodyPiece w3 = new WurmBodyPiece(theControl, w2, head);
                w3.EatenByParasite +=new EventHandler<EventArgs>(PieceEatenByParasite);
                theControl.AddCreature(w3, starting_pos);
                w2 = w3;
            }
            tail = w2;
            head.Eats += new EventHandler<EatEventData>(head_Eats);
        }

        void PieceEatenByParasite(object sender, EventArgs e)
        {
            WurmBodyPiece w = (WurmBodyPiece)sender;
            if (w.Follower != null)
            {
                w.Follower.Leader = w.Leader;
            }
            if (w == tail)
                tail = w.Leader;

            theControl.RemovePiece(w);
            length--;
            if (LengthChange != null)
                LengthChange(this, new GameEventArgs(theControl));
        }

        void head_Eats(object sender, EatEventData e)
        {
            if (!e.CreatureEaten.Eaten)
            {
                food_eaten_since_growth += e.CreatureEaten.Size;
                e.CreatureEaten.Eat();
                if (Eats != null)
                    Eats(this, new GameEventArgs(theControl));
            }
            while (food_eaten_since_growth >= food_per_segment)
            {
                food_eaten_since_growth -= food_per_segment;
                WurmBodyPiece wp = new WurmBodyPiece(theControl, tail, head);
                wp.EatenByParasite +=new EventHandler<EventArgs>(PieceEatenByParasite);
                theControl.AddCreature(wp, tail.CenterPoint);
                tail = wp;
                length++;
                if (LengthChange != null)
                    LengthChange(this, new GameEventArgs(theControl));
            }
        }

        public int Length
        {
            get { return length; }
        }
    }
}

File: ts/wurm.ts
//import { relWurmHeadRotate } from "./timing.js";
import {AddCreature, RemovePiece} from "./gameControl.js";
import {BackAttachable, WurmBodyPiece, WurmHead} from "./wurmPieces.js";

const food_per_segment = 4;

export class Wurm {
    length: number;
    head : WurmHead;
    tail : BackAttachable;

    food_eaten_since_growth : number =0;

    constructor(length:number, startX:number, startY:number){
        this.length=length;
        this.head = new WurmHead(this);
        AddCreature(this.head, startX, startY);

        let w2 = new WurmBodyPiece(this.head,this.head);
        //parasite handler
        AddCreature(w2, startX, startY);

        for (let  i = 0; i < length-2; i++){
            let w3 = new WurmBodyPiece(w2, this.head);
            //parasite handler
            AddCreature(w3, startX, startY);
            w2 = w3;
        }
        this.tail = w2;
        //head eats handler
    }

    get Length() :number{
        return this.length;
    }

    pieceEatenByParasite(w:WurmBodyPiece){
        if (w.Follower != null)
        {
            w.Follower.Leader = w.Leader;
        }
        if (w == this.tail)
            this.tail = w.Leader;

        RemovePiece(w);
        this.length--;
    }

    head_Eats(head : WurmHead) :void
        {
            if (!head.target.eaten)
            {
                this.food_eaten_since_growth += head.target.Eat();
                //if (Eats != null)
                //    Eats(this, new GameEventArgs(theControl));
            }
            while (this.food_eaten_since_growth >= food_per_segment)
            {
                this.food_eaten_since_growth -= food_per_segment;
                let wp = new WurmBodyPiece(this.tail, head);
                //wp.EatenByParasite +=new EventHandler<EventArgs>(PieceEatenByParasite);
                AddCreature(wp, this.tail.backAttachX, this.tail.backAttachY);
                this.tail = wp;
                this.length++;
                //if (LengthChange != null)
                //    LengthChange(this, new GameEventArgs(theControl));
            }
        }
    
}